action/protocol/vote/epochmetapb/epochmeta.proto

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) 2020 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+// To compile the proto, run:
+//      protoc --go_out=plugins=grpc:. *.proto
+
+syntax = "proto3";
+package epochmetapb;
+
+import "google/protobuf/timestamp.proto";
+
+message EpochMeta {
+  uint64 epochNumber = 1;
+  repeated BlockMeta blockMetas = 2;
```

Comment:
what's your plan of "EpochMeta"?
I can see that you may want to define more fields in the struct.I have two ideas:
1. shall they be grouped together? For example, you can have a list of length 720 * E to store the last 720 * E block meta, which including block producer, block time, block hash and etc. So, it may not be appropriate to be placed in epoch meta. I am not saying grouping by block is a better way, just be sure you already compared them
2. why do we have epochNumber field? we may want to store last E epoch meta info and they can be stored in an array (fixed size) with a pointer to current epoch. In this case, if we know the current epoch number and the pointer, we can tell the epoch number of each meta in the array. Moreover, in this way, we don't need to rotate (write a large field) or write multiple fields in the state db.

Reply:
I couldn't fully understand about what you are saying. Just to confirm my understanding, 
1. "Group-by-block" means not having epoch concept as epochmeta does, but just storing latest 720 block meta information, is that correct? And just retrieve current productivity from start to end, by comparing blockmeta's height, right? But then we have to shift to append latest block, right? Currently, we don't have to shift but only append or initialize to be a new struct at epoch start height. 
2. We only have to store current epoch meta for this now productivity usage. I can remove epoch number field because we don't need it. What do you mean by having a pointer to get an epoch number? You are saying is that for example, having a fixed size array(720) and having two array indexer(epoch start point and erase point) and replace the oldest block meta? 

If my understanding is correct, in order to satisfy current usage, because current productivity is calculated on epoch unit, just having only current epoch's block meta and initialize those at epoch start height don't need shifting and indexing, so it looks fine. I can get rid of epochNumber field in epochmeta (since we are not using), IMO it's okay to keep having current epoch's block meta for current usage. But do you think will there be other usages that require keeping latest 720 blocks? If so, I can change as "Group-by-block" way for future extension.

Reply:
thanks for the details, now I have a full picture what you want to do. Initially, I thought you want to store multiple epoch metas.

about block meta part, storing metas of the latest 720 blocks by 720 state db keys.
  pros:
      a. This will minimize writes to state db. To tell what's the next block meta to write to, we need an extra index in state db, NEXT_BLOCK_META_INDEX. To append, we just need to write the block meta into state db by key 'BLOCK_META_KEY_' + value_in_state_db(NEXT_BLOCK_META_INDEX).
      b. Benefit other usage of block meta, say in consensus, we need to get block time and etc.
  cons:
      a. Significantly increase the read cost if we cannot read in batch.